File: e2e/tests/worktree.test.ts
/**
 * E2E Tests: Worktree Feature
 *
 * Tests the git worktree mode functionality which allows creating
 * isolated development sessions in separate git worktrees.
 *
 * These tests require a crush.json with worktree_mode: true in the
 * e2e directory.
 */

import { test, expect } from "@microsoft/tui-test";
import * as path from "path";
import * as os from "os";

// Resolve binary path
const CRUSH_BINARY =
  process.env.CRUSH_BINARY || path.resolve(process.cwd(), "../crush");

// Helper to wait for TUI to be ready - needs enough time for init dialog
const STARTUP_DELAY = 3000;

// Skip git repo tests on Windows (bash script)
const isWindows = os.platform() === "win32";

/**
 * Worktree Dialog Tests
 *
 * Tests the worktree creation dialog that appears when pressing ctrl+n
 * with worktree_mode enabled.
 *
 * Note: These tests run from the e2e directory which has crush.json
 * with worktree_mode: true configured.
 */
test.describe("Worktree dialog", () => {
  test.use({
    program: {
      file: CRUSH_BINARY,
      args: [],
    },
  });

  test("opens on ctrl+n", async ({ terminal }) => {
    // Wait for TUI to initialize (project init dialog appears first)
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));

    // Press ctrl+n to open new session dialog
    terminal.write("\x0e"); // ctrl+n = ASCII 14

    // Should show the worktree dialog
    await expect(
      terminal.getByText("New Worktree Session", { full: true })
    ).toBeVisible();
  });

  test("shows dialog description", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");

    await expect(
      terminal.getByText("Enter a name for the git worktree branch", {
        full: true,
      })
    ).toBeVisible();
  });

  test("shows session name label", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");

    await expect(
      terminal.getByText("Session name:", { full: true })
    ).toBeVisible();
  });

  test("shows placeholder text", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");

    await expect(
      terminal.getByText("feature-name", { full: true })
    ).toBeVisible();
  });

  test("shows keybinding hints", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");

    await expect(
      terminal.getByText("create worktree", { full: true })
    ).toBeVisible();
    await expect(terminal.getByText("cancel", { full: true })).toBeVisible();
  });

  test("closes dialog on escape", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");

    // Verify dialog is open
    await expect(
      terminal.getByText("New Worktree Session", { full: true })
    ).toBeVisible();

    // Press escape to close
    terminal.keyEscape();
    await new Promise((r) => setTimeout(r, 500));

    // Dialog should be closed - check that project init dialog is back
    await expect(
      terminal.getByText("initialize this project", { full: true })
    ).toBeVisible();
  });

  test("shows error for empty name on submit", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");
    await new Promise((r) => setTimeout(r, 300));

    // Press enter without typing a name
    terminal.submit();
    await new Promise((r) => setTimeout(r, 300));

    // Should show error message
    await expect(
      terminal.getByText("Name is required", { full: true })
    ).toBeVisible();
  });

  test("accepts input in text field", async ({ terminal }) => {
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");
    await new Promise((r) => setTimeout(r, 300));

    // Type a branch name
    terminal.write("my-test-feature");
    await new Promise((r) => setTimeout(r, 200));

    // Should show typed text
    await expect(
      terminal.getByText("my-test-feature", { full: true })
    ).toBeVisible();
  });
});

/**
 * Worktree Creation Tests (in temp git repo)
 *
 * These tests create a temporary git repository to test actual worktree
 * creation. Uses a bash wrapper that:
 * 1. Creates a temp directory
 * 2. Initializes a git repo with an initial commit
 * 3. Creates crush.json with worktree_mode enabled
 * 4. Runs crush in that directory
 */
test.describe("Worktree creation in git repo", () => {
  // Build the bash script with the binary path embedded
  const setupAndRunCrush = `
set -e
TMPDIR=$(mktemp -d)
trap 'rm -rf "$TMPDIR"' EXIT
cd "$TMPDIR"
git init -q
git config user.email "test@test.com"
git config user.name "Test User"
echo "# Test Project" > README.md
git add README.md
git commit -q -m "Initial commit"
cat > crush.json << 'CRUSHCONFIG'
{
  "options": {
    "worktree_mode": true
  }
}
CRUSHCONFIG
exec "${CRUSH_BINARY}"
`;

  test.use({
    program: {
      file: "bash",
      args: ["-c", setupAndRunCrush.replace("${CRUSH_BINARY}", CRUSH_BINARY)],
    },
  });

  test.when(
    !isWindows,
    "creates worktree with valid name",
    async ({ terminal }) => {
      // Wait for TUI to initialize
      await new Promise((r) => setTimeout(r, STARTUP_DELAY));

      // Press ctrl+n to open worktree dialog
      terminal.write("\x0e");
      await new Promise((r) => setTimeout(r, 500));

      // Verify dialog opened
      await expect(
        terminal.getByText("New Worktree Session", { full: true })
      ).toBeVisible();

      // Type a valid branch name
      const branchName = `test-feature-${Date.now()}`;
      terminal.write(branchName);
      await new Promise((r) => setTimeout(r, 200));

      // Submit the form
      terminal.submit();
      await new Promise((r) => setTimeout(r, 2000));

      // Should show success message
      await expect(
        terminal.getByText("Worktree created", { full: true })
      ).toBeVisible();
    }
  );

  test.when(
    !isWindows,
    "sanitizes special characters in branch name",
    async ({ terminal }) => {
      await new Promise((r) => setTimeout(r, STARTUP_DELAY));
      terminal.write("\x0e");
      await new Promise((r) => setTimeout(r, 500));

      // Type a name with spaces (will be sanitized to hyphens)
      terminal.write("my new feature");
      await new Promise((r) => setTimeout(r, 200));

      terminal.submit();
      await new Promise((r) => setTimeout(r, 2000));

      // Should succeed (name gets sanitized)
      await expect(
        terminal.getByText("Worktree created", { full: true })
      ).toBeVisible();
    }
  );
});

/**
 * Worktree Error Handling Tests
 *
 * Tests error conditions when worktree creation fails.
 */
test.describe("Worktree error handling", () => {
  test.use({
    program: {
      file: CRUSH_BINARY,
      args: [],
    },
  });

  test("shows error when not in git repo", async ({ terminal }) => {
    // When creating worktree from e2e/ (not a git repo root), should show error
    await new Promise((r) => setTimeout(r, STARTUP_DELAY));
    terminal.write("\x0e");
    await new Promise((r) => setTimeout(r, 300));

    terminal.write("test-feature");
    await new Promise((r) => setTimeout(r, 200));

    terminal.submit();
    await new Promise((r) => setTimeout(r, 1000));

    // Should show error about not being a git repository
    await expect(
      terminal.getByText("not a git repository", { full: true })
    ).toBeVisible();
  });
});

File: e2e-next/worktree_test.go
package e2e

import (
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/aleksclark/trifle"
)

const (
	startupDelay     = 3 * time.Second
	dialogTransition = 500 * time.Millisecond
)

// TestWorktreeDialogOpens tests that the worktree dialog opens with ctrl+n.
func TestWorktreeDialogOpens(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	// Wait for TUI to initialize.
	time.Sleep(startupDelay)

	// Press ctrl+n to open new session dialog.
	if err := term.Write(trifle.CtrlN()); err != nil {
		t.Fatalf("Failed to write: %v", err)
	}

	// Should show the worktree dialog.
	locator := term.GetByText("New Worktree Session", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected worktree dialog: %v", err)
	}
}

// TestWorktreeDialogDescription tests the dialog description.
func TestWorktreeDialogDescription(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())

	locator := term.GetByText("Enter a name for the git worktree branch", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected dialog description: %v", err)
	}
}

// TestWorktreeDialogLabel tests the session name label.
func TestWorktreeDialogLabel(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())

	locator := term.GetByText("Session name:", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected session name label: %v", err)
	}
}

// TestWorktreeDialogPlaceholder tests the placeholder text.
func TestWorktreeDialogPlaceholder(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())

	locator := term.GetByText("feature-name", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected placeholder: %v", err)
	}
}

// TestWorktreeDialogKeybindings tests that keybinding hints are shown.
func TestWorktreeDialogKeybindings(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())

	if err := term.GetByText("create worktree", trifle.WithFull()).WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected create keybinding: %v", err)
	}
	if err := term.GetByText("cancel", trifle.WithFull()).WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected cancel keybinding: %v", err)
	}
}

// TestWorktreeDialogClose tests that escape closes the dialog.
func TestWorktreeDialogClose(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())

	// Verify dialog is open.
	if err := term.GetByText("New Worktree Session", trifle.WithFull()).WaitVisible(5 * time.Second); err != nil {
		t.Fatalf("Dialog did not open: %v", err)
	}

	// Press escape to close.
	_ = term.KeyEscape()
	time.Sleep(dialogTransition)

	// Check that project init dialog is back.
	locator := term.GetByText("initialize this project", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected to return to init dialog: %v", err)
	}
}

// TestWorktreeDialogEmptyError tests error for empty name.
func TestWorktreeDialogEmptyError(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())
	time.Sleep(300 * time.Millisecond)

	// Press enter without typing a name.
	_ = term.Submit()
	time.Sleep(300 * time.Millisecond)

	// Should show error message.
	locator := term.GetByText("Name is required", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected error message: %v", err)
	}
}

// TestWorktreeDialogInput tests text input acceptance.
func TestWorktreeDialogInput(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())
	time.Sleep(300 * time.Millisecond)

	// Type a branch name.
	_ = term.Write("my-test-feature")
	time.Sleep(200 * time.Millisecond)

	// Should show typed text.
	locator := term.GetByText("my-test-feature", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected typed text: %v", err)
	}
}

// TestWorktreeCreationInGitRepo tests worktree creation in a temp git repo.
func TestWorktreeCreationInGitRepo(t *testing.T) {
	trifle.SkipOnWindows(t)

	// Bash script that creates a temp git repo and runs crush.
	script := fmt.Sprintf(`
set -e
TMPDIR=$(mktemp -d)
trap 'rm -rf "$TMPDIR"' EXIT
cd "$TMPDIR"
git init -q
git config user.email "test@test.com"
git config user.name "Test User"
echo "# Test Project" > README.md
git add README.md
git commit -q -m "Initial commit"
cat > crush.json << 'CRUSHCONFIG'
{
  "options": {
    "worktree_mode": true
  }
}
CRUSHCONFIG
exec "%s"
`, CrushBinary())

	term, err := trifle.NewTerminal("bash", []string{"-c", script}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)

	// Open worktree dialog.
	_ = term.Write(trifle.CtrlN())
	time.Sleep(dialogTransition)

	// Verify dialog opened.
	if err := term.GetByText("New Worktree Session", trifle.WithFull()).WaitVisible(5 * time.Second); err != nil {
		t.Fatalf("Dialog did not open: %v", err)
	}

	// Type a valid branch name.
	branchName := fmt.Sprintf("test-feature-%d", time.Now().UnixNano())
	_ = term.Write(branchName)
	time.Sleep(200 * time.Millisecond)

	// Submit the form.
	_ = term.Submit()
	time.Sleep(2 * time.Second)

	// Should show success message.
	locator := term.GetByText("Worktree created", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected success message: %v", err)
	}
}

// TestWorktreeSanitization tests that special characters are sanitized.
func TestWorktreeSanitization(t *testing.T) {
	trifle.SkipOnWindows(t)

	script := fmt.Sprintf(`
set -e
TMPDIR=$(mktemp -d)
trap 'rm -rf "$TMPDIR"' EXIT
cd "$TMPDIR"
git init -q
git config user.email "test@test.com"
git config user.name "Test User"
echo "# Test Project" > README.md
git add README.md
git commit -q -m "Initial commit"
cat > crush.json << 'CRUSHCONFIG'
{
  "options": {
    "worktree_mode": true
  }
}
CRUSHCONFIG
exec "%s"
`, CrushBinary())

	term, err := trifle.NewTerminal("bash", []string{"-c", script}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())
	time.Sleep(dialogTransition)

	// Type a name with spaces (will be sanitized to hyphens).
	_ = term.Write("my new feature")
	time.Sleep(200 * time.Millisecond)

	_ = term.Submit()
	time.Sleep(2 * time.Second)

	// Should succeed (name gets sanitized).
	locator := term.GetByText("Worktree created", trifle.WithFull())
	if err := locator.WaitVisible(5 * time.Second); err != nil {
		t.Errorf("Expected success after sanitization: %v", err)
	}
}

// TestWorktreeNotGitRepo tests error when not in git repo.
func TestWorktreeNotGitRepo(t *testing.T) {
	trifle.SkipOnWindows(t)

	term, err := trifle.NewTerminal(CrushBinary(), []string{}, trifle.TerminalOptions{
		Rows: 40,
		Cols: 100,
	})
	if err != nil {
		t.Fatalf("Failed to create terminal: %v", err)
	}
	defer term.Close()

	time.Sleep(startupDelay)
	_ = term.Write(trifle.CtrlN())
	time.Sleep(300 * time.Millisecond)

	_ = term.Write("test-feature")
	time.Sleep(200 * time.Millisecond)

	_ = term.Submit()
	time.Sleep(1 * time.Second)

	// Should show error about not being a git repository.
	output := term.Output()
	if !strings.Contains(output, "not a git repository") {
		t.Errorf("Expected git repo error, got: %s", output)
	}
}
